### PR TITLE
Classifier chain variants + test fixes

### PR DESCRIPTION
**Pull request by jacobmontiel**
This PR includes two variants of the classifier chains:

- `ProbabilisticClassifierChain`
- `MonteCarloClassifierChain`

The PR also includes minor bug fixes that result in errors during the CI tests.

### CODE REVIEW

File: river/multioutput/chain.py [hunk 1]
```diff
@@ -231,8 +252,199 @@ def predict_one(self, x):
...
+    >>> dataset = synth.Logical(seed=42, n_tiles=100)
+
+    >>> model = multioutput.ProbabilisticClassifierChain(
+    ...     model=linear_model.LogisticRegression()
+    ... )
+
+    >>> metric = metrics.Jaccard()
+
+    >>> for x, y in dataset:
+    ...    # Convert y values to booleans
```
Comment (smastelini) [inline river/multioutput/chain.py:301]: Is this comment valid?

File: river/multioutput/chain.py [hunk 2]
```diff
@@ -231,8 +252,199 @@ def predict_one(self, x):
...
+    >>> model = multioutput.MonteCarloClassifierChain(
+    ...     model=linear_model.LogisticRegression(),
+    ...     m=10,
+    ...     seed=42
+    ... )
+
+    >>> metric = metrics.Jaccard()
+
+    >>> for x, y in dataset:
+    ...    # Convert y values to booleans
```
Comment (smastelini) [inline river/multioutput/chain.py:397]: Same here

**Comment by smastelini** [comment]
Hi @jacobmontiel, great work. One suggestion: what about replacing the usage of `iter_sklearn` in the examples by the datasets `Music` and `SolarFlare` (for multi-label classification and multi-target regression, respectively)?

File: river/multioutput/chain.py [hunk 3]
```diff
@@ -231,8 +252,199 @@ def predict_one(self, x):
...
             y_pred[o] = clf.predict_one(x)
             x[o] = y_pred[o]
 
         return y_pred
+
+
+class ProbabilisticClassifierChain(ClassifierChain):
+    r"""Probabilistic Classifier Chains.
+
+    The Probabilistic Classifier Chains (PCC) [1^] is a Bayes-optimal method
```
Comment (smastelini) [inline river/multioutput/chain.py:268]:
```suggestion
    The Probabilistic Classifier Chains (PCC) [^1] is a Bayes-optimal method
```

File: river/multioutput/chain.py [hunk 4]
```diff
@@ -231,8 +252,199 @@ def predict_one(self, x):
...
+        return prod(p.values())
+
+
+class MonteCarloClassifierChain(ProbabilisticClassifierChain):
+    """Monte Carlo Sampling Classifier Chains.
+
+    Probabilistic Classifier Chains using Monte Carlo sampling, as
+    described in [^1].
+
+    `m samples are taken from the posterior distribution. Therefore we
```
Comment (smastelini) [inline river/multioutput/chain.py:365]:
```suggestion
    m samples are taken from the posterior distribution. Therefore we
```

File: river/multioutput/chain.py [hunk 5]
```diff
@@ -231,8 +252,199 @@ def predict_one(self, x):
...
+
+        max_payoff = 0.
+        n_labels = len(self.order)
+        # for each and every possible label combination
+        for label in range(2 ** n_labels):
+            # put together a binary label vector
+            y_gen = {i: int(v) for i, v in zip(self.order, list(bin(label)[2:].zfill(n_labels)))}
+            # ... and gauge a probability for it (given x)
+            payoff = self._payoff(x=x, y=y_gen)
+            # print(label, y_gen, payoff)
```
Comment (smastelini) [inline river/multioutput/chain.py:335]: what about removing it?

**Review by smastelini — CHANGES_REQUESTED** [review]
Hi @jacobmontiel, great work. I finished a first pass in the code.

One suggestion: what about replacing the usage of iter_sklearn in the examples by the datasets Music and SolarFlare (for multi-label classification and multi-target regression, respectively)?

File: river/multioutput/chain.py (hunk 1 above)
Comment (jacobmontiel) [reply river/multioutput/chain.py:301]: No, it was unintentionally left. I removed it.

File: river/multioutput/chain.py (hunk 2 above)
Comment (jacobmontiel) [reply river/multioutput/chain.py:397]: Same.

File: river/multioutput/chain.py (hunk 5 above)
Comment (jacobmontiel) [reply river/multioutput/chain.py:335]: What a noob. I am removing this.

**Comment by jacobmontiel** [comment]
> Hi @jacobmontiel, great work. I finished a first pass in the code.
> 
> One suggestion: what about replacing the usage of iter_sklearn in the examples by the datasets Music and SolarFlare (for multi-label classification and multi-target regression, respectively)?

Using `Music` is straightforward:
```
    >>> from river import feature_selection
    >>> from river import linear_model
    >>> from river import metrics
    >>> from river import multioutput
    >>> from river import preprocessing
    >>> from river import datasets

    >>> dataset = datasets.Music()
    >>> model = preprocessing.StandardScaler()
    >>> model |= multioutput.ClassifierChain(
    ...    model=linear_model.LogisticRegression(),
    ...    order=None
    ... )

    >>> metric = metrics.Jaccard()

    >>> for x, y in dataset:
    ...     y_pred = model.predict_one(x)
    ...     metric = metric.update(y, y_pred)
    ...     model = model.learn_one(x, y)

    >>> metric
    Jaccard: 0.503069
```

However, I like that the example shows how to use external data sets. On the other hand, `SolarFlare` is a little more complicated as it requires additional steps (transform `str` to `int`). We could update the SolarFlare class to set an option to return numeric data directly. But that is out of the scope of this PR.

**Comment by smastelini** [comment]
> However, I like that the example shows how to use external data sets. 

That's indeed a good point!

**Review by smastelini — COMMENTED** [review]
I finished a second pass and everything looks good to me! Are all tests with `pytest` passing? If so, I will merge this PR. :)

**Comment by jacobmontiel** [comment]
```= 1210 passed, 114 deselected, 4 warnings in 103.95s (0:01:43) =```

**Comment by MaxHalford** [comment]
The pull request is [here](https://travis-ci.org/github/creme-ml/creme/builds/738331977) on Travis. I have to set it up so that it shows up on GitHub!

Have a great weekend.

**Comment by smastelini** [comment]
Hi @jacobmontiel,

Travis yields this error:

image

I also got this when running on my local (linux) machine. The result in the docstring doesn't match what is generated.

**Comment by jacobmontiel** [comment]
All tests are passing again!

**Comment by MaxHalford** [comment]
LGTM!

**Pull request merged**